developer-docs-site/static/move_demo/sources/CAPMessageBoard.move

```diff
@@ -0,0 +1,202 @@
+        let board = borrow_global<CapMessageBoard>(board_addr);
+        board.notice
+    }
+
+    /// participant register to init his/her resource
+    public fun register_internal(participant: &signer) {
+        move_to(participant, NoticeChangeEventHandle{
+            change_events: Event::new_event_handle<NoticeChangeEvent>(participant)
+        })
+    }
+
+    public(script) fun register(participant: signer) {
+        register_internal(&participant);
+    }
+
+    /// board owner control adding new participants
+    public(script) fun add_participant(account: signer, participant: address) acquires NoticeCapEventHandle {
+        add_participant_internal(&account, participant);
+    }
+
+    public fun add_participant_internal(account: &signer, participant: address) acquires NoticeCapEventHandle {
+        let board_addr = Signer::address_of(account);
+        Offer::create(account, NoticeChangeCapability{ board: board_addr }, participant);
+
+        let event_handle = borrow_global_mut<NoticeCapEventHandle>(board_addr);
+
+        Event::emit_event<NoticeCapEvent>(
+            &mut event_handle.change_events,
+            NoticeCapEvent{
+                participant,
+                board: board_addr
+            }
+        );
+    }
+
+    public(script) fun claim_notice_cap(account: signer, board: address) {
+        claim_notice_cap_internal(&account, board);
+    }
+
+    /// claim offered notice capability
+    public fun claim_notice_cap_internal(account: &signer, board: address) {
+        let notice_cap = Offer::redeem<NoticeChangeCapability>(
+            account, board);
+        move_to(account, notice_cap);
+    }
+
+    /// remove a participant capability to publish notice
+    public(script) fun remove_participant(account: signer, participant: address) acquires NoticeChangeCapability {
+        let cap = borrow_global_mut<NoticeChangeCapability>(participant);
+        assert!(Signer::address_of(&account) == cap.board, EONLY_ADMIN_CAN_REMOVE_NOTICE_CAP);
+        let cap = move_from<NoticeChangeCapability>(participant);
+        let NoticeChangeCapability{ board: _ } = cap;
+    }
+
+    /// only the participant with right capability can publish the message
+    public fun submit_pinned_message_internal(
+        cap: &NoticeChangeCapability, sender: address, board_addr: address, message: vector<u8>
+    ) acquires NoticeChangeEventHandle, CapMessageBoard {
+        assert!(cap.board == board_addr, EACCOUNT_NO_NOTICE_CAP);
+        let board = borrow_global_mut<CapMessageBoard>(board_addr);
+        board.notice = message;
+        let event_handle = borrow_global_mut<NoticeChangeEventHandle>(sender);
+        Event::emit_event<NoticeChangeEvent>(
+            &mut event_handle.change_events,
+            NoticeChangeEvent{
+                sender,
+                message,
+                board: board_addr
+            }
+        );
+    }
+
+    public(script) fun submit_pinned_message(
```

Comment:
shouldn't we have a submit message that just creates events?

Reply:
It is not very intuitive for me to have a submit message function for a message board, which is more like a functionality provided by the chatroom. 

we also demonstrate creating event when submitted pinned_message. not sure what additional feature can be demonstrated by submit_message

Reply:
I was thinking we'd have the events represent the chatroom, since storing them on-chain is expensive and not the right type of construct we want to support.

What are you suggesting represents the chatroom?

Reply:
updated and added the function just creating events